### ISSUE TITLE
Ability to "declare" stubs as functions

### ISSUE THREAD

**Issue by ezzatron** [eloquent/phony#163]
There's probably a good opportunity to replace usage of [isolator](https://github.com/IcecaveStudios/isolator) with a strategy similar to [php-mock](https://github.com/php-mock/php-mock). That is, declaring functions in a namespace to exploit PHP's function resolution rules, and to swap out real functions with stubbed and spied implementations.

I'm thinking something like this:

```php
namespace My\Namespace;

function rollDice()
{
    return mt_rand(1, 6);
}

$stub = stub('mt_rand');
$stub->with(1, 6)->returns(6);
$stub->declareAs('My\Namespace\mt_rand');

$result = rollDice();

assert(6 === $result);
$stub->calledWith(1, 6);
```

In the above example, `My\Namespace\mt_rand` is declared, but its implementation can later be replaced by any stub that declares itself with the same name. That may take some interesting trickery.

Thoughts @jmalloc, @darianbr, @koden-km, @parkertr?

**Comment by ezzatron** [comment]
Could probably add a `stubGlobal()` convenience method that also makes the stub return `null` by default, a bit like a full mock (stubs don't default to this behavior, only mocks):

```php
namespace My\Namespace;

use function Eloquent\Phony\stubGlobal;

$stub = stubGlobal('file_get_contents', __NAMESPACE__);

assert(null === file_get_contents('/path/to/secrets'));
$stub->calledWith('/path/to/secrets');
```

**Comment by parkertr** [comment]
I'm sure of any technical nuances here but I very much like this feature.

**Comment by jmalloc** [comment]
Erin, notes as per discussion on "stub scope criteria" below :) It would not be necessary to implement any of this as the initial offering, if pursued at all.

### Rationale

While, IMO, this system described in this issue is superior to an isolator, especially in terms of performance, there is some degree of flexibility lost.  We are currently able to create multiple isolators with separate behaviour in each instance. This sometimes becomes necessary when tests require multiple instances of the SUT.

By allowing stubs to match on the scope of the caller we can give them differing behaviour depending on where they are invoked.  This is especially useful with the "global stubs" feature where we have limited control over the (re)definition of the global PHP functions.

### Stub criteria

All stubs would gain a method for adding a "calling scope" criteria.  If I have the terminology correct, this criteria limits the following stub answers to only those calls made inside a stack frame with a matching scope.  The scope could be a global function, a static or instance method, a specific class, or a specific instance of a class.

Such a criteria would be defined something like the following:

`$stub->withCaller($functionOrClassOrInstance, $methodName = null)->does(...)`

I suspect it wont be necessary to resolve the meaning of the first parameter ahead of time, rather this could be deferred until the calling stack frame (obtained via `debug_backtrace`) is inspected.

### Global stubbing convenience parameters

The `stubGlobal` function described in the comment above would gain the criteria parameters:

`stubGlobal($fn, $namespace, $functionOrClassOrInstance = null, $methodName = null)`

If these parameters are omitted the function would behave as described above. If at least `$functionOrClassOrInstance` is provided, the "global stub" would instead dispatch to a "child" stub only when this criteria is met.  This "child" stub would be returned by `stubGlobal`, allowing the user to independently configure the behaviour without having to be aware of the scope criteria.

By way of example:
```php
function stubGlobal($fn, $namespace, $functionOrClassOrInstance = null, $methodName = null) {
    $stub = /* create stub using "existing" behaviour */

    // there is no scope criteria, return the global stub
    if ($functionOrClassOrInstance === null) {
        return $stub;
    }

    // wire the global stub to dispatch to a child stub when the criteria is matched
    $child = Phony::stub($fn);
    $stub->withCaller($functionOrClassOrInstance, $methodName)->does($child);

    return $child;
}
```

### In summary

... yeah it's madness.

**Comment by ezzatron** [comment]
I'm in favour of the `withCaller()` idea in general, but I think it's probably better to accept standard PHP callables in a single parameter, rather than `$functionOrClassOrInstance` and `$methodName`:

```php
$stub->withCaller($closure);
$stub->withCaller('functionName');
$stub->withCaller('ClassName::methodName');
$stub->withCaller(['ClassName', 'methodName']);
$stub->withCaller([$instance, 'methodName']);

$stub = stubGlobal('file_get_contents', __NAMESPACE__, [$instance, 'methodName']);
```

Particularly since Phony already deals with these kinds of callable definitions in other places. The only question then is the semantics of how to indicate that *any* method from a particular instance or class should be matched. Perhaps an asterisk:

```php
$stub->withCaller([$instance, '*']);
$stub = stubGlobal('file_get_contents', __NAMESPACE__, [$instance, '*']);
```

?

**Comment by jmalloc** [comment]
I'm not opposed to that, but one thing to consider is that they aren't *quite* `callable`, so you wont be able to type hint as such. Specifically, when you want to refer to a non-static method, but match *any* instance, could might refer to this as `'ClassName::nonStaticMethod'` or `['ClassName', 'nonStaticMethod']`, but that would be confused for a static method by PHP.  Hopefully that makes sense.

**Comment by ezzatron** [comment]
Can't type hint anyway - 5.3 support.

**Comment by jmalloc** [comment]
I LOL'D.

**Comment by jmalloc** [comment]
Oops, I didn't see your part about `[$instance, '*']`, I figured you'd just pass the instance itself, but that discounts closures and other invocables, so yep I'm on board with that.

**Comment by ezzatron** [comment]
I just realised that by-reference arguments make this more complicated. I'm going to have to define the "hook" function with the same parameters as the underlying stub in order to get them working. That means you can't switch out the implementation unless the function signatures are compatible.

Unfortunately I feel that it's a must-have. Many functions that are prime candidates for this feature take arguments by reference (i.e. `exec()`'s `$output` parameter).
